sdk/core/core/src/az_span_seq.c

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_span_seq.h>
+
+#include <az_write_span_iter.h>
+#include <az_str.h>
+
+#include <stdlib.h>
+
+#include <_az_cfg.h>
+
+//
+
+AZ_CALLBACK_DATA(az_span_span_callback, az_span_span const *, az_span_seq)
+
+az_result az_span_span_to_seq_func(
+    az_span_span const * const context,
+    az_span_visitor const visitor) {
+  AZ_CONTRACT_ARG_NOT_NULL(context);
+
+  size_t const size = context->size;
+  az_const_span const * begin = context->begin;
+  for (size_t i = 0; i < size; ++i) {
+    AZ_RETURN_IF_FAILED(visitor.func(visitor.data, begin[i]));
+  }
+  return AZ_OK;
+}
+
+az_span_seq az_span_span_to_seq(az_span_span const * const p_span) {
+  return az_span_span_callback(p_span, az_span_span_to_seq_func);
+}
+
+//
+
+AZ_CALLBACK_DATA(az_size_callback, size_t *, az_span_visitor)
+
+az_result az_span_add_size(size_t * const p_size, az_const_span const span) {
+  *p_size += span.size;
+  return AZ_OK;
+}
+
+az_result az_span_seq_size(az_span_seq const seq, size_t * const out_size) {
+  AZ_CONTRACT_ARG_NOT_NULL(out_size);
+
+  *out_size = 0;
+  return seq.func(seq.data, az_size_callback(out_size, az_span_add_size));
+}
+
+//
+
+az_result az_span_seq_to_str(az_span_seq const seq, az_span const span) {
+  az_write_span_iter i = az_write_span_iter_create(span);
+  AZ_RETURN_IF_FAILED(seq.func(seq.data, az_write_span_iter_to_span_visitor(&i)));
+  AZ_RETURN_IF_FAILED(az_write_span_iter_write(&i, AZ_STR("\0")));
+  return AZ_OK;
+}
+
+//
+
+az_result az_span_seq_to_new_str(az_span_seq const seq, char * * const out) { 
+  *out = NULL;
+  size_t size; 
+  AZ_RETURN_IF_FAILED(az_span_seq_size(seq, &size));
+  size += 1;
+  uint8_t * const p = (uint8_t *)malloc(size);
+  if (p == NULL) {
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
```

Comment:
The SDK is non-allocating.

If this is needed for CURL transport layer then move it into that header to avoid the impression that it can be used everywhere.

Reply:
I agree that we need clear separation between allocating and non-allocating utility functions, but I'm not sure that allocating utility functions should go to curl transport layer.

Reply:
I'm Ok to move to CURL .C file for now.